src/git-source-provider.ts

```diff
@@ -109,8 +110,20 @@ export async function getSource(settings: IGitSourceSettings): Promise<void> {
 
     // Fetch
     core.startGroup('Fetching the repository')
-    const refSpec = refHelper.getRefSpec(settings.ref, settings.commit)
-    await git.fetch(settings.fetchDepth, refSpec)
+    if (settings.fetchDepth <= 0) {
+      let refSpec = refHelper.getRefSpecForAllHistory(
+        settings.ref,
+        settings.commit
+      )
+      await git.fetch(refSpec)
+      if (!(await refHelper.testRef(git, settings.ref, settings.commit))) {
```

Comment:
When all history is fetched (all branches and all tags), the ref we're interested in may have moved to a different commit (push or force push)

Test whether the ref points to the expected commit. If not, then fetch again with a targeted refspec.

I'll add some comments in the code here.